src/core/Target.ml

```diff
@@ -12,18 +12,52 @@
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
  * LICENSE for more details.
  *)
-(*****************************************************************************)
-(* Prelude *)
-(*****************************************************************************)
-(* A target file.
- *
- * See AST_generic.program for more information. This module just
- * defines an alias to have better naming convetion to differentiate
- * a target program from a pattern program.
- *)
 
-(*****************************************************************************)
-(* Types *)
-(*****************************************************************************)
+(* See Target.mli for documentation of public items. *)
+
+type path = { origin : Origin.t; internal_path_to_content : Fpath.t }
+[@@deriving show, eq]
+
+type manifest = { path : path; kind : Manifest_kind.t } [@@deriving show]
+
+type lockfile = {
+  path : path;
+  kind : Lockfile_kind.t;
+  manifest : manifest option;
+}
+[@@deriving show]
+
+type regular = {
+  path : path;
+  analyzer : Xlang.t;
+  products : Semgrep_output_v1_t.product list;
+  lockfile : lockfile option;
+}
+[@@deriving show]
+
+type t = Regular of regular | Lockfile of lockfile [@@deriving show]
```

Comment:
In theory because both variants have this path field, you could move it up to factorize.
But maybe it's fine like this.

Reply:
There are some other threads with similar suggestions, but I do not think that works since we need to be able to distinguish later between regular and lockfile targets. If this were instead
```ocaml
type u = Regular of regular | Lockfile of lockfile (* without path field for either *)
type t = { path : path; target : u }
```
then if I am passing around a value of `Target.regular` I also need to separately keep the path, and if I use a value of type `t` then I cannot enforce at the type level it is particularly a regular or lockfile target.

Something like
```ocaml
type 'a t = { path : path; target : 'a }
type regular = regular_info t;
type lockfile = lockfile_info t;
type any = (regular_info, lockfile_info) Either.t t;
```
could work, but Martin was of the opinion such a pattern should be avoided due to error messages & inference.